gnomad_constraint/utils/constraint.py

```diff
@@ -386,3 +392,171 @@ def apply_models(
+        # Filter to LoF annotations with LOFTEE HC or OS.
+        "lof_hc_os": (ht.modifier == "HC") | (ht.modifier == "OS"),
+        # Filter to LoF annotations with LOFTEE HC.
+        "lof": ht.modifier == "HC",
+        # Filter to missense variants.
+        "mis": ht.annotation == "missense_variant",
+        # Filter to probably damaging missense variants predicted by PolyPen-2.
+        "mis_pphen": ht.modifier == "probably_damaging",
+        # Filter to synonymous variants.
+        "syn": ht.annotation == "synonymous_variant",
+    }
+
+    # Compute the observed:expected ratio. Will not compute per pop for "mis_pphen".
+    ht = ht.group_by(*keys).aggregate(
+        **{
+            ann: oe_aggregation_expr(
+                ht,
+                filter_expr,
+                pops=() if ann == "mis_pphen" else pops,
+                exclude_mu_sum=True if ann == "mis_pphen" else False,
+            )
+            for ann, filter_expr in annotation_dict.items()
+        }
+    )
+    ht = ht.checkpoint(
+        new_temp_file(prefix="compute_constraint_metrics", extension="ht")
+    )
+
+    # Compute the pLI scores for LoF variants.
+    ht = ht.annotate(
+        **{
+            ann: ht[ann].annotate(**compute_pli(ht, ht[ann].obs, ht[ann].exp))
+            for ann in ["lof_hc_lc", "lof_hc_os", "lof"]
+        }
+    )
+
+    # Compute the 90% confidence interval around the observed:expected ratio and z
+    # scores.
+    ht = ht.annotate(
+        **{
+            ann: ht[ann].annotate(
+                oe_ci=oe_confidence_interval(ht[ann].obs, ht[ann].exp),
+                z_raw=calculate_raw_z_score(ht[ann].obs, ht[ann].exp),
+            )
+            for ann in ["lof", "mis", "syn"]
+        }
+    )
+
+    ht = ht.annotate(
+        **{
+            ann: ht[ann].annotate(
+                constraint_flags=add_constraint_flags(
+                    exp_expr=ht[ann].exp,
+                    outlier_expr=ht[ann].z_raw < -raw_z_outlier_threshold
+                    if ann != "syn"
+                    else hl.abs(ht[ann].z_raw) > raw_z_outlier_threshold,
+                )
+            )
+            for ann in ["lof", "mis", "syn"]
+        }
+    )
+
+    # Add 'no_variants' to the constraint flags if there are a total of 0 observed
+    # variants summed across lof, mis, and syn.
+    ht = ht.annotate(
+        **{
+            ann: ht[ann].annotate(
+                constraint_flags=hl.if_else(
+                    hl.or_else(ht.lof["obs"], 0)
+                    + hl.or_else(ht.mis["obs"], 0)
+                    + hl.or_else(ht.syn["obs"], 0)
+                    == 0,
+                    ht[ann].constraint_flags.add("no_variants"),
+                    ht[ann].constraint_flags,
+                )
+            )
+            for ann in ["lof", "mis", "syn"]
+        }
+    )
+
+    ht = ht.annotate(
+        **{
+            ann: ht[ann].annotate(
+                **calculate_z_score(
```

Comment:
I changed this to only compute the aggregated sds, add them to globals, and then compute the z-score within this code since it's a simple divide, but we could also add that simple divide to a `calculate_z_score` if you think that would be better